### ISSUE TITLE
Strange case of empty non-coalesced sparse tensor

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#73479]
### 🐛 Describe the bug

[bug.pt.gz](https://github.com/pytorch/pytorch/files/8147613/bug.pt.gz)

1. I don't understand why a.indices() gives an error (maybe related https://github.com/pytorch/pytorch/issues/73426)
2. If indices() of empty tensor is supposed to fail, `why torch.zeros(...).indices()` works?
3. Why difference of behavior of `torch.zeros_like(...).indices()` and `torch.zeros(...).indices()`?

Thanks!

```python
a = torch.load('bug.pt')
print(a)
#tensor(indices=tensor([], size=(3, 0)),
#       values=tensor([], size=(0,)),
#       size=(12, 1, 1), nnz=0, layout=torch.sparse_coo)

print(a.indices())
# RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first

print(torch.zeros_like(a))
#tensor(indices=tensor([], size=(3, 0)),
#       values=tensor([], size=(0,)),
#       size=(12, 1, 1), nnz=0, layout=torch.sparse_coo)

print(torch.zeros_like(a).indices())
# RuntimeError: Cannot get indices on an uncoalesced tensor, please call .coalesce() first

print(torch.zeros(12, 1, 1, layout = torch.sparse_coo, dtype = a.dtype, device = a.device).indices())
# tensor([], size=(3, 0), dtype=torch.int64)
```

### Versions

1.9.1

**Comment by cpuhrsch** [comment]
@vadimkantorov - does this hold true on the latest nightly?

**Comment by vadimkantorov** [comment]
haven't checked - will check in two weeks...

**Comment by pearu** [comment]
To answer some of the "why" questions here: In general, operators on sparse tensors return a coaleased or uncoalesced sparse tensor. Which one is returned, depends on the operator and if the operator is able to detect/ensure/guarantee that the constructed sparse tensor is truly coalesced. It is preferable for an operator to return a coalesced tensor but only if no coalesce op is required. Otherwise, the caller needs to coalesce the tensor explicitly if it needs to call `indicies()` or `values()` methods. As evident from this issue, not all operators try to return a coalesced tensor even though the return sparse tensor is truly coalesced. I'd consider this discussion as an optimization matter rather than a bug.

On the other hand, I think the underlying problem of this issue lies in `torch.save` function that does not store the coalesceness state of sparse tensors. For instance:
```python
>>> s=torch.tensor([[1, 0], [3, 4]]).to_sparse().coalesce()
>>> s.is_coalesced()
True
>>> torch.save(s, 'sparse.pt')
>>> s1 = torch.load('sparse.pt')
>>> s1.is_coalesced()
False
```
So, basically, whenever loading a tensor from `.pt` file, one would need to coalesce it explicitly if the coalesced version of the tensor is required.

**Comment by vadimkantorov** [comment]
> On the other hand, I think the underlying problem of this issue lies in `torch.save` function that does not store the coalesceness state of sparse tensors. For instance:

I think I had the same results in live debugging session. Tensor `a` was uncoalesced even before saving.

I think zeros_like should always produced a coalesced-marked tensor, especially given that nnz=0.

I also think that to_sparse() and maybe sparse_coo constructor should also mark a tensor as coalesced if nnz=0, regardless of what operator has produced it.

I guess preserving of coalesced-ness is a separate bug of torch.load/torch.save. Should I create the corresponding issues? Or will you do it?

**Comment by pearu** [comment]
@vadimkantorov If you don't mind creating an issue for torch.save/load vs sparse coo tensor discrepancy, then that would be great!

**Comment by ezyang** [comment]
B is now fixed on master, but the others are still broken

**Comment by vadimkantorov** [comment]
hmm, it's strange that C is not fixed

**Comment by ezyang** [comment]
C doesn't need to be fixed, it always worked 📦

**Comment by vadimkantorov** [comment]
Then I guess the remaining things are: 
- supporting empty indices for empty tensor
- serialization to preserve coalesced-ness
- public binding for nnz (would be more explicit and choice-free than `indices().numel()`/`values().numel()`)

**Comment by vadimkantorov** [comment]
@pearu what's left in this issue then?

* serialization to preserve coalesced-ness
* public binding to obtain number of sparse elements

?

**Comment by pearu** [comment]
With the current master we have:
```python
>>> import torch
>>> a = torch.zeros((12, 1, 1), layout=torch.sparse_coo)
>>> a
tensor(indices=tensor([], size=(3, 0)),
       values=tensor([], size=(0,)),
       size=(12, 1, 1), nnz=0, layout=torch.sparse_coo)
>>> print(a.indices())
tensor([], size=(3, 0), dtype=torch.int64)
>>> print(a.is_coalesced())
True
>>> b = torch.zeros_like(a)
>>> print(b)
tensor(indices=tensor([], size=(3, 0)),
       values=tensor([], size=(0,)),
       size=(12, 1, 1), nnz=0, layout=torch.sparse_coo)
>>> print(b.indices())
tensor([], size=(3, 0), dtype=torch.int64)
>>> print(b.is_coalesced())
True
>>> c = torch.zeros(a.shape, layout = a.layout, dtype = a.dtype, device = a.device)
>>> print(c.indices())
tensor([], size=(3, 0), dtype=torch.int64)
>>> print(c.is_coalesced())
True
>>> d= torch.tensor([]).to_sparse()
>>> print(d.indices())
tensor([], size=(1, 0), dtype=torch.int64)
>>> print(d.is_coalesced())
True
```
that is, most examples work as expected, including save/load an empty coo tensor.

What does not work is saving/loading a coalesced tensor as in https://github.com/pytorch/pytorch/issues/73479#issuecomment-1056859335. This corresponds to the "serialization to preserve coalesced-ness" issue. I suggest the following solution:
- saving always coalesces COO tensors so that loading can flag the COO tensors as coalesced by default
- reasoning: the uncoalesce feature is in most use cases an intermediate optimization step of creating a COO tensor to minimize expensive index sorting tasks when the COO tensor is updated via adding elements to it. Saving a COO tensor is about persisting the tensor to be used possibly from another Python process. The suggestion is about assuming that the processes of creating COO tensors and of saving-loading the COO tensors, do not overlap.
- a possible BC issue: users having truly uncoalesced COO tensors stored will experience undefined behavior when loading these tensors with the coalesced flag set to true. Considering the beta state of sparse support, would this be an acceptable BC breakage case?

Re "public binding to obtain number of sparse elements", < rant > I have always advocated using `nse` instead of `nnz` (starting from #45695 [not merged], [RFC-0004, item 14](https://github.com/Quansight-Labs/rfcs/tree/pearu/rfc-fill-value/RFC-0004-sparse-fill-value#future-extensions-and-existing-issues), and ending with using NSE in documentations and discussions where appropriate) but it is hard to break habits of using NNZ even when people know that this is a misleading acronym. It is unfortunate that people entering this field will be educated to use the acronym until they discover the hard way that it is actually misleading.. </ rant > So, I agree with the issue but I am not sure it can be resolved within this issue.

**Comment by vadimkantorov** [comment]
e.g. when saving a narrow'd big tensor, torch.save still saves the full tensor without cloning - i would argue that this is even more confusing than doing coalesce before saving. but there should be some way of doing de/serialization that preserves the data structures exactly - for debugging and reporting bugs, this is super useful

**Comment by vadimkantorov** [comment]
i guess nnz has maybe some history as well? probably because it was a matlab function: https://www.mathworks.com/help/matlab/ref/nnz.html

**Comment by pearu** [comment]
`nnz` in matlab is totally fine as it is defined as `(tensor != 0).sum()`. For non-hybrid sparse tensors, that would be `(tensor.values() != 0).sum()`. The problem is in sparse formats history when `nnz` was introduced as `tensor.values().size(0)`.

Currently, `torch.Tensor._nnz` is a private method that is defined as `tensor._values().size(0)` and it can be variable: typically the value reduces after `coalesce()` call. Its use ought to be internal only.

I think we don't need to refer to `torch.Tensor._nnz` or change its semantics. Instead, request a new public feature that includes methods that compute various statistics on sparse tensors:
- sparse tensor storage efficiency defined as `tensor.values().numel()` (perhaps plus the overhead of storing indices) and that can be compared to the strided tensor storage efficiency defined as `tensor.numel()` (warning: 0-strides and tensor slices may distort the meaning of `numel()` result!)
- sparse tensor storage inefficiency defined as `tensor.values() == 0` that would represent theoretical storage efficiency of the sparse tensor
- anything else?

**Comment by pearu** [comment]
> e.g. when saving a narrow'd big tensor, torch.save still saves the full tensor without cloning - i would argue that this is even more confusing 

I agree.

> than doing coalesce before saving. but there should be some way of doing de/serialization that preserves the data structures exactly - for debugging and reporting bugs, this is super useful

To preserve the data structures exactly, I guess this boils down to de/serialization of storage object and supporting recreation of the data structure using a given storage object.

**Comment by vadimkantorov** [comment]
about nnz: i think both physical values.numel() is useful (for various tricks, e.g. checking if the tensor is empty / actually has any elements or not) and semantic one checking that it (values == 0).sum()

**Comment by themaigod** [comment]
Any updates for requesting a new public feature, instead of _nnz()? 
I can find nnz for matrix in scipy.sparse, and also like downstream library DGL. So it is necessary to align with common processing, especailly when people can see this attribute when print the sparse tensor object.
Maybe people have doubt on the definition of such a function as current discussions. Just make it public as a temp solution is better than do nothing.
